rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+   ];
+}
+
+// DatabaseObjectImportRuleSpec is the spec for database object import rule.
+message DatabaseObjectImportRuleSpec {
+  // Priority represents the priority of the rule application. Lower numbered rules will be applied first.
+  int32 Priority = 1 [(gogoproto.jsontag) = "priority"];
+
+  // DatabaseLabels is a set of labels which must match the database for the rule to be applied.
+  map<string, string> DatabaseLabels = 2 [(gogoproto.jsontag) = "db_labels,omitempty"];
+
+  // Mappings is a list of matches that will map match conditions to labels.
+  repeated DatabaseObjectImportRuleMapping Mappings = 3 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "mappings,omitempty"
+  ];
+}
+```
+
+Individual objects are matched with against all object matches defined in
+`DatabaseObjectImportRuleMapping` message. If any of the matches succeeds (or if
+the list of matches is empty), the labels specified in the mapping are applied
+to the object. The processing continues with the next mapping: multiple mappings
+can match any given object.
+
+```protobuf
+// DatabaseObjectImportRuleMapping is the mapping between object properties and labels that will be added to the object.
+message DatabaseObjectImportRuleMapping {
+  // ObjectMatches is a set of object matching rules for this mapping.
+  // For a given database object, each of the matches is attempted.
+  // If any of them succeed, the labels are applied.
+  repeated DatabaseObjectSpec ObjectMatches = 2 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "object_match"
+  ];
+
+  // AddLabels specifies which labels to add if any of the previous matches match.
+  map<string, string> AddLabels = 3 [(gogoproto.jsontag) = "add_labels"];
+}
+```
+
+##### Resource: `db_object`
+
+Another new resource is the database object (`db_object`), imported using the
+import rules from the database.
+
+The spec for this resource is the `DatabaseObjectSpec` message, which is also
+used in the `DatabaseObjectImportRuleMapping`.
+
+The spec is equivalent to a `map<string, string>`, except it predefines a few
+optional properties.
+
+In case of an `db_object`, the entirety of the spec is provided by
+database-specific schema introspection tool. The custom attributes provide a way
+to express additional properties important to the object, which cannot be
+rightly placed in other attributes. The `object_kind` property is mandatory for
+`db_object` resources.
+
+```protobuf
+// DatabaseObjectSpec is the spec for the database object.
+message DatabaseObjectSpec {
+   string Protocol = 1 [(gogoproto.jsontag) = "protocol,omitempty"];
+   string ServiceName = 2 [(gogoproto.jsontag) = "service_name,omitempty"];
+   string ObjectKind = 3 [(gogoproto.jsontag) = "object_kind"];
+   string Database = 4 [(gogoproto.jsontag) = "database,omitempty"];
+   string Schema = 5 [(gogoproto.jsontag) = "schema,omitempty"];
+   string Name = 6 [(gogoproto.jsontag) = "name,omitempty"];
```

Comment:
nit `Name`, `Protocol`, `ServiceName`, `ObjectKind` shouldn't be empty?

Reply:
`omitempty` is purely a hint for pretty printing, so it only affects the UX. For the database object resource the validation of required fields is done separately from these annotations. For writing the matching expressions, the UX matters: a bunch of empty fields add a lot of noise, hence why I only left the `object_kind` as a single field to be shown as default. 

I'm not 100% sure this is a right call, though, but the important bit is that the perspective here is the UX in writing (and reading) the matchers.

Reply:
it's really a nit thing. When I am reading this, to me at least, `omitempty` implies the field is optional.

Reply:
I've added `omitempty` to all of those, now.